68_JpegLoading/main.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2018-2024 - DevSH GrapMonoAssetManagerAndBuiltinResourceApplicationhics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#include "nbl/application_templates/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+
+#include <thread>
+#include <atomic>
+#include <mutex>
+#include <queue>
+#include <functional>
+#include <condition_variable>
+
+#include "nlohmann/json.hpp"
+#include "argparse/argparse.hpp"
+
+using json = nlohmann::json;
+
+using namespace nbl;
+using namespace core;
+using namespace hlsl;
+using namespace system;
+using namespace asset;
+using namespace ui;
+using namespace video;
+
+class ThreadPool {
+public:
+	using task_t = std::function<void()>;
+
+	ThreadPool() {
+		for (size_t i = 0; i < m_maxWorkers; i++) {
+			m_threads.emplace_back([this]() {
+				while (true) {
+					task_t task;
+
+					{
+						std::unique_lock<std::mutex> lock(m_queueMutex);
+						m_taskAvailable.wait(lock, [&]() { return !m_tasks.empty() || m_stop; });
+
+						if (m_stop && m_tasks.empty())
+							return;
+
+						task = std::move(m_tasks.front());
+						m_tasks.pop();
+					}
+
+					task();
+				}
+				});
+		}
+	}
+
+	~ThreadPool() {
+		{
+			std::lock_guard<std::mutex> lock(m_queueMutex);
+			m_stop = true;
+		}
+
+		m_taskAvailable.notify_all();
+
+		for (auto& thread : m_threads) {
+			thread.join();
+		}
+	}
+
+	void addTask(task_t task) {
+		std::lock_guard<std::mutex> queueLock(m_queueMutex);
+		m_tasks.emplace(std::move(task));
+		m_taskAvailable.notify_one();
+	}
+private:
+	const size_t m_maxWorkers = std::thread::hardware_concurrency();
+	std::vector<std::thread> m_threads;
+	std::condition_variable m_taskAvailable;
+	bool m_stop = false; // stops the threads on object destruction
```

Comment:
should be an atomic_flag (or something that supports memory semantic), a plain boolean may get its loads and stores eliminated

Reply:
stop flag is now atomic [8ad3d18](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/161/commits/8ad3d18fc785b5b3e92a57e983c0135f9beaf3ae)